### PR TITLE
fix(memory): add qmd get fallback for get memory read file

### PR DESCRIPTION
**Pull request by velisiska**
## Summary

- Problem: When a qmd document path is slugified (e.g. by the qmd index), the file doesn't exist at the expected filesystem location and `readFile` returns empty text
- Why it matters: Agents reading memory docs get blank content for valid qmd documents with slugified paths
- What changed: Added a `qmd get` fallback in `readFile` — when a `qmd/` file is missing on disk, it converts the path to a `qmd://` URI and delegates to `qmd get` which can resolve slugified paths. Extracted `getQmdFile` as a clean helper that accepts a ready URI
- What did NOT change (scope boundary): Path-traversal protections (`resolveReadPath` collection allow-list + `isWithinRoot`) still run before the fallback; no changes to write paths or search

## Change Type (select all)

- [x] Bug fix

## Scope (select all touched areas)

- [x] Memory / storage

## Linked Issue/PR

- Related: n/a

## User-visible / Behavior Changes

Memory `readFile` now returns document content for slugified qmd paths instead of empty text.

## Security Impact (required)

- New permissions/capabilities? No
- Secrets/tokens handling changed? No
- New/changed network calls? No
- Command/tool execution surface changed? No — `qmd get` is already available; this adds a new call site bounded by existing collection allow-list and `isWithinRoot` checks
- Data access scope changed? No

### Steps

1. Create a qmd collection with a document that gets a slugified path
2. Call `readFile` with the slugified `qmd/<collection>/<slug>.md` path
3. Before fix: returns `{ text: "" }`; after fix: returns the document content

### Expected

- Document content returned via `qmd get` fallback

### Actual (before fix)

- Empty text returned

## Evidence

- [x] Failing test/log before + passing after
- 3 new tests: happy-path fallback, `from`/`lines` slicing, `qmd get` failure gracefully returns empty text

## Human Verification (required)

- Verified scenarios: all 3 new tests pass (`pnpm test -- src/memory/qmd-manager.test.ts`)
- Edge cases checked: `qmd get` failure returns empty text (not error); `from`/`lines` slicing works on fallback content
- What you did **not** verify: end-to-end with a real `qmd` binary (mocked in tests)

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

## Compatibility / Migration

- Backward compatible? Yes
- Config/env changes? No
- Migration needed? No

## Failure Recovery (if this breaks)

- How to disable/revert this change quickly: revert the commit on `src/memory/qmd-manager.ts`
- Known bad symptoms reviewers should watch for: unexpected `qmd get` spawns for non-slugified paths (should not happen — fallback only triggers when file is missing on disk)

## Risks and Mitigations

- Risk: Extra `qmd get` spawn for genuinely missing files (not slugified, just nonexistent)
  - Mitigation: `getQmdFile` returns `null` on failure; falls through to existing empty-text return. Cost is one bounded subprocess call with the configured timeout.

### CODE REVIEW

File: src/memory/qmd-manager.ts [hunk 1]
```diff
@@ -893,6 +893,20 @@ export class QmdMemoryManager implements MemorySearchManager {
...
+      // Fall back to `qmd get` which can resolve slugified paths.
+      if (relPath.startsWith("qmd/")) {
+        const qmdFile = relPath.replace(/^qmd\//, "qmd://");
+        const qmdText = await this.getQmdFile(qmdFile);
+        if (qmdText !== null) {
+          if (params.from !== undefined || params.lines !== undefined) {
+            const allLines = qmdText.split("\n");
+            const start = Math.max(1, params.from ?? 1);
+            const count = Math.max(1, params.lines ?? allLines.length);
+            return { text: allLines.slice(start - 1, start - 1 + count).join("\n"), path: relPath };
```
[the remark this answers is not in the document]
Comment (velisiska) [reply src/memory/qmd-manager.ts:905]: This is consistent with the existing `split("\n")` fallback at lines 912–916 (for full files with `from`/`lines`), which has the same trailing-blank-line behavior. The `readPartialText` path (using `readline`) behaves differently, but that's a pre-existing inconsistency — not introduced by this change. No action needed.

File: src/memory/qmd-manager.ts [hunk 2]
```diff
@@ -1765,6 +1779,17 @@ export class QmdMemoryManager implements MemorySearchManager {
     return !path.isAbsolute(relativePath);
   }
 
+  private async getQmdFile(file: string): Promise<string | null> {
+    try {
+      const result = await this.runQmd(["get", file], {
+        timeoutMs: this.qmd.limits.timeoutMs,
+      });
+      return result.stdout ?? null;
+    } catch {
```
[the remark this answers is not in the document]
Comment (velisiska) [reply src/memory/qmd-manager.ts:1788]: A 200k+ markdown document in memory context is an extreme edge case. The graceful degradation to `{ text: "" }` (same as missing) is acceptable here — the normal file-read path has analogous size constraints (files are read fully into memory). If this ever becomes a real issue we can revisit, but adding complexity for this edge case isn't warranted now.

**Comment by vincentkoc** [comment]
Thanks @velisiska. Closing as superseded. Current main now keeps QMD paths roundtrip-safe and resolves slugified  results back to readable OpenClaw paths through the indexed-path mapping path and slugified-path regression coverage, so this older  fallback branch is no longer the right merge vehicle. The active fixes landed in the recent slugified-path / roundtrip-safe merge train on main.

**Comment by vincentkoc** [comment]
Clarifying the closure note: current main already resolves slugified `qmd://...` search hits back to readable OpenClaw paths through the indexed-path mapping path and roundtrip-safe read-file coverage, so this older `qmd get` fallback branch is superseded by the canonical slugified-path / roundtrip-safe fixes now on main.
